strategies/categorized/src/test/java/org/arquillian/smart/testing/strategies/categorized/CategorizedTestsDetectorTest.java

```diff
@@ -0,0 +1,137 @@
+
+        // when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(classesToProcess);
+
+        // then
+        assertThat(testSelection).hasSize(2);
+        assertThat(SmartTesting.getClasses(new HashSet<>(testSelection)))
+            .contains(FirstAndSecondCategorizedClass.class, FirstCategorizedClass.class);
+    }
+
+    @Test
+    public void should_return_class_with_both_first_and_second_category_using_match_all() {
+        // given
+        categorizedConfig.setCategories(
+            new String[] {FirstCategory.class.getSimpleName(), SecondCategory.class.getSimpleName().toLowerCase()});
+        categorizedConfig.setMatchAll(true);
+
+        // when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(classesToProcess);
+
+        // then
+        assertThat(testSelection).hasSize(1);
+        assertThat(SmartTesting.getClasses(new HashSet<>(testSelection)))
+            .contains(FirstAndSecondCategorizedClass.class);
+    }
+
+    @Test
+    public void should_return_class_with_second_category_when_case_sensitivity_is_set() {
+        // given
+        categorizedConfig.setCategories(
+            new String[] {FirstCategory.class.getSimpleName().toLowerCase(), SecondCategory.class.getSimpleName()});
+        categorizedConfig.setCaseSensitive(true);
+
+        // when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(classesToProcess);
+
+        // then
+        assertThat(testSelection).hasSize(1);
+        assertThat(SmartTesting.getClasses(new HashSet<>(testSelection)))
+            .contains(FirstAndSecondCategorizedClass.class);
+    }
+
+    @Test
+    public void should_return_all_classes_with_any_category_when_no_category_is_set() {
+        // when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(classesToProcess);
+
+        // then
+        assertThat(testSelection).hasSize(3);
+        assertThat(SmartTesting.getClasses(new HashSet<>(testSelection)))
+            .contains(ThirdCategorizedClass.class, FirstAndSecondCategorizedClass.class, FirstCategorizedClass.class);
+    }
+
+    @Test
+    public void should_return_classes_without_first_category_when_reversed_is_used() {
+        // given
+        categorizedConfig.setCategories(
+            new String[] {FirstCategory.class.getSimpleName()});
+        categorizedConfig.setReversed(true);
+
+        // when
+        Collection<TestSelection> testSelection =
+            new CategorizedTestsDetector(config).selectTestsFromClasses(classesToProcess);
+
+        // then
+        assertThat(testSelection).hasSize(2);
+        assertThat(SmartTesting.getClasses(new HashSet<>(testSelection)))
+            .contains(ThirdCategorizedClass.class, NonCategorizedClass.class);
```

Comment:
Instead of calling size maybe we can change those assertions to one using `containsExactly(...)`?

Reply:
Yeah, and in case of multiple classes: `containsExactlyInAnyOrder` so I don't have to care about the order :-)